src/core/checkpoints/index.ts

```
@@ -153,11 +164,12 @@ async function checkGitInstallation(
 		})
 
 		log("[Task#getCheckpointService] initializing shadow git")
-
-		service.initShadowGit().catch((err) => {
+		try {
+			await service.initShadowGit()
```

We’ve seen that this can be very slow in large codebases, and leads to a bad experience where it seems like nothing happens after typing the first message of a task. Is there some way to keep it async but ensure it finishes before any write operation?

---

The effect you mentioned has been achieved. Because the initialization method is called asynchronously.

src\core\task\Task.ts:1297-1299
```
private async initiateTaskLoop(userContent: Anthropic.Messages.ContentBlockParam[]): Promise<void> {
		// Kicks off the checkpoints initialization process in the background.
		getCheckpointService(this)
```

It's just that the logic of initialization before writing was not adjusted at that time

I removed the getInitializedCheckpointService method and made getCheckpointService wait, while checkpointSave will call the getCheckpointService method synchronously, thus ensuring that it waits for the checkpoint to be initialized before the write operation is performed.

---

Oh nice! Sorry, reviewing on my phone.

---

That seems like the right tradeoff!

---

sounds like that may also fix the issue where "Initial Checkpoint" overrides/hides an error message.